### PR TITLE
Add Cardano transactions proofs support to mithril-client

### LINKED ISSUE #1468: Implement transactions verification in client library
## Why
The client library needs to provide a way to generate, verify and certify proofs for a list of transactions.

## What
Implement new API function for verifying a list of proofs generated by the aggregator in the client library.

## How

- [x] Implement transaction verification function in library:
  - [x] Call aggregator to generate a list of proofs for the transaction list
  - [x] Return the list of certified transactions (and list of not certified?)
  - [x] Verify the proofs for all the certified transactions 
  - [x] Retrieve the latest artifact (and its certificate hash)
  - [x] Verify that the Merkle root of the Merkle proofs is signed by the certificate
  - [x] Verify that the certificate chain is valid
- [x] Implement a function that lists artifacts or get latest produced artifact?
- [x] ~~Update Mithril events feedback mechanism~~
- [x] Update documentation
- [x] Add feature flag `unstable`?

## Later
- Add an example in the repository (later, will be done in another PR)

### PR DESCRIPTION
**Pull request by Alenar**
## Content

This PR add Cardano transactions proofs support to the `mithril-client` library.
The added capabilities are:
- List and get CardanoTransactions Commitments
- Get a verifiable proofs that a list of transactions hashes are part of the Cardano chain.

Since they're still work in progress those capabilities are gated behind a `unstable` feature.

## Pre-submit checklist

- Branch
  - [x] Tests are provided (if possible)
  - [x] Crates versions are updated (if relevant)
  - [x] Commit sequence broadly makes sense
  - [x] Key commits have useful messages
- PR
  - [x] No clippy warnings in the CI
  - [x] Self-reviewed the diff
  - [x] Useful pull request description
  - [x] Reviewer requested
- Documentation
  - [ ] Update README file (if relevant)
  - [ ] Update documentation website (if relevant)
  - [ ] Add dev blog post (if relevant)

## Comments
Should the `verify` method of the `CardanoTransactionsSetProof` entity fail if its `transactions_hashes` list is empty ? It's should be an "impossible" case but it's constructor allow it.

## Issue(s)
Closes #1468

### CODE REVIEW

**Review by dlachaumepalo — APPROVED** [review]

File: mithril-client/tests/cardano_transaction_proof.rs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        Some(format!(
+            "/{}",
+            AggregatorRequest::GetCertificate {
+                hash: proofs.certificate_hash.clone()
+            }
+            .route()
+        ))
+    );
+
+    // 4 - validate that the verified transactions match the signed by the certificate
```
Comment (jpraynaud) [inline mithril-client/tests/cardano_transaction_proof.rs:58]:
```suggestion
    // 4 - validate that the verified transactions proof is signed by the certificate
```

File: mithril-common/src/messages/cardano_transaction_commitment.rs [hunk 2]
```diff
@@ -72,7 +72,7 @@ mod tests {
             "certificate_hash": "certificate-hash-123",
             "created_at": "2023-01-19T13:43:05.618857482Z"
         }"#;
-        let message: CardanoTransactionMessage = serde_json::from_str(json).expect(
+        let message: CardanoTransactionCommitmentMessage = serde_json::from_str(json).expect(
             "This JSON is expected to be succesfully parsed into a CardanoTransactionMessage instance.",
```
Comment (jpraynaud) [inline mithril-common/src/messages/cardano_transaction_commitment.rs:76]:
```suggestion
            "This JSON is expected to be succesfully parsed into a CardanoTransactionCommitmentMessage instance.",
```

File: mithril-common/src/messages/cardano_transaction_commitment_list.rs [hunk 3]
```diff
@@ -78,7 +78,7 @@ mod tests {
         }]"#;
         println!("message: {:?}", golden_message());
 
-        let message: CardanoTransactionListMessage = serde_json::from_str(json).expect(
+        let message: CardanoTransactionCommitmentListMessage = serde_json::from_str(json).expect(
                     "This JSON is expected to be succesfully parsed into a CardanoTransactionListMessage instance.",
```
Comment (jpraynaud) [inline mithril-common/src/messages/cardano_transaction_commitment_list.rs:82]:
```suggestion
                    "This JSON is expected to be succesfully parsed into a CardanoTransactionCommitmentListMessage instance.",
```

File: mithril-common/src/messages/cardano_transactions_proof.rs [hunk 4]
```diff
@@ -1,18 +1,80 @@
...
 
     /// Transactions that could not be certified
-    non_certified_transactions: Vec<TransactionHash>,
+    pub non_certified_transactions: Vec<TransactionHash>,
+}
+
+/// Set of transactions verified by [CardanoTransactionsProofsMessage::verify].
+///
+/// Can be used to reconstruct part of a [ProtocolMessage] in order to check that
+/// it is indeed associated to the certificate it claims to.
```
Comment (jpraynaud) [inline mithril-common/src/messages/cardano_transactions_proof.rs:25]:
```suggestion
/// it is indeed signed by a certificate.
```

File: mithril-common/src/messages/cardano_transactions_proof.rs [hunk 5]
```diff
@@ -1,18 +1,80 @@
...
+#[derive(Debug, Clone, PartialEq)]
+pub struct VerifiedCardanoTransactions {
+    certificate_hash: String,
+    merkle_root: String,
+    certified_transactions: Vec<TransactionHash>,
+}
+
+impl VerifiedCardanoTransactions {
+    /// Hash of the certificate that sign this struct merkle root.
```
Comment (jpraynaud) [inline mithril-common/src/messages/cardano_transactions_proof.rs:34]:
```suggestion
    /// Hash of the certificate that signs this struct Merkle root.
```

File: mithril-common/src/messages/cardano_transactions_proof.rs [hunk 6]
```diff
@@ -1,18 +1,80 @@
...
+pub enum VerifyCardanoTransactionsProofsError {
+    /// The verification of an individual [CardanoTransactionsSetProofMessagePart] failed.
+    #[error("Invalid set proof for transactions hashes: {transactions_hashes:?}")]
+    InvalidSetProof {
+        transactions_hashes: Vec<TransactionHash>,
+        source: StdError,
+    },
+
+    /// No certified transactions set proof to verify
+    #[error("There's no certified transactions set proof to verify")]
```
Comment (jpraynaud) [inline mithril-common/src/messages/cardano_transactions_proof.rs:64]: Maybe like this?
```suggestion
    #[error("There's no certified transaction to verify")]
```

File: mithril-common/src/messages/cardano_transactions_proof.rs [hunk 7]
```diff
@@ -1,18 +1,80 @@
...
+    NoCertifiedTransactions,
```
Comment (jpraynaud) [inline mithril-common/src/messages/cardano_transactions_proof.rs:65]:
```suggestion
    NoCertifiedTransaction,
```

File: mithril-common/src/messages/cardano_transactions_proof.rs [hunk 8]
```diff
@@ -1,18 +1,80 @@
...
+
+    /// Not all certified transactions set proof have the same merkle root.
+    ///
+    /// This is problematic because all the set proof should be generated from the same
+    /// merkle tree whose root is signed in the [crate::entities::Certificate][certificate].
```
Comment (jpraynaud) [inline mithril-common/src/messages/cardano_transactions_proof.rs:70]:
```suggestion
    /// merkle tree which root is signed in the [crate::entities::Certificate][certificate].
```

File: mithril-common/src/messages/cardano_transactions_proof.rs [hunk 9]
```diff
@@ -1,18 +1,80 @@
...
+    #[error("All certified transactions set proof must share the same merkle root")]
```
Comment (jpraynaud) [inline mithril-common/src/messages/cardano_transactions_proof.rs:71]:
```suggestion
    #[error("All certified transactions set proofs must share the same Merkle root")]
```

File: mithril-common/src/messages/cardano_transactions_proof.rs [hunk 10]
```diff
@@ -28,4 +90,179 @@ impl CardanoTransactionsProofsMessage {
...
     }
+
+    /// Verify that all the certified transactions proofs are valid
+    ///
+    /// The following checks will be executed:
+    /// 1 - Check for each merkle proof that its leaves belongs to their merkle root
+    /// 2 - Check that all proofs merkle roots are the same
+    /// 3 - Assert that there's at least one certified transaction
+    ///
+    /// If every check is okay, the hex encoded merkle root of the proof will be returned.
```
Comment (jpraynaud) [inline mithril-common/src/messages/cardano_transactions_proof.rs:101]:
```suggestion
    /// The following checks will be executed:
    /// 1 - Check that each Merkle proof is valid
    /// 2 - Check that all proofs share the same Merkle root
    /// 3 - Assert that there's at least one certified transaction
    ///
    /// If every check is okay, the hex encoded Merkle root of the proof will be returned.
```

File: openapi.yaml [hunk 11]
```diff
@@ -1476,7 +1476,7 @@ components:
         "protocol_parameters": { "k": 5, "m": 100, "phi_f": 0.65 }
         }
 
-    CardanoTransactionListMessage:
+    CardanoTransactionCommitmentListMessage:
       description: CardanoTransactionListMessage represents a list of Cardano transactions set
```
Comment (jpraynaud) [inline openapi.yaml:1480]:
```suggestion
      description: CardanoTransactionCommitmentListMessage represents a list of Cardano transactions set commitments
```

File: openapi.yaml [hunk 12]
```diff
@@ -1520,7 +1520,7 @@ components:
             "created_at": "2022-06-14T10:52:31Z"
           }
 
-    CardanoTransactionMessage:
+    CardanoTransactionCommitmentMessage:
       description: This message represents a Cardano transactions set.
```
Comment (jpraynaud) [inline openapi.yaml:1524]:
```suggestion
      description: This message represents a Cardano transactions set commitment.
```

File: openapi.yaml [hunk 13]
```diff
@@ -267,52 +267,52 @@ paths:
 
   /artifact/cardano-transactions:
     get:
-      summary: Get most recent Cardano transactions sets
+      summary: Get most recent Cardano transactions commitments
```
Comment (jpraynaud) [inline openapi.yaml:270]:
```suggestion
      summary: Get most recent Cardano transactions set commitments
```

File: openapi.yaml [hunk 14]
```diff
@@ -267,52 +267,52 @@ paths:
...
       description: |
-        Returns the list of the most recent Cardano transactions sets
+        Returns the list of the most recent Cardano transactions commitments
```
Comment (jpraynaud) [inline openapi.yaml:272]:
```suggestion
        Returns the list of the most recent Cardano transactions set commitments
```

File: openapi.yaml [hunk 15]
```diff
@@ -267,52 +267,52 @@ paths:
...
       responses:
         "200":
-          description: Cardano transactions set found
+          description: Cardano transactions commitments found
```
Comment (jpraynaud) [inline openapi.yaml:275]:
```suggestion
          description: Cardano transactions set commitments found
```

File: openapi.yaml [hunk 16]
```diff
@@ -267,52 +267,52 @@ paths:
...
           content:
             application/json:
               schema:
-                $ref: "#/components/schemas/CardanoTransactionListMessage"
+                $ref: "#/components/schemas/CardanoTransactionCommitmentListMessage"
         "412":
           description: API version mismatch
         default:
-          description: Cardano transactions set retrieval error
+          description: Cardano transactions commitments retrieval error
```
Comment (jpraynaud) [inline openapi.yaml:283]:
```suggestion
          description: Cardano transactions set commitments retrieval error
```

File: openapi.yaml [hunk 17]
```diff
@@ -267,52 +267,52 @@ paths:
...
           content:
             application/json:
               schema:
                 $ref: "#/components/schemas/Error"
 
   /artifact/cardano-transaction/{hash}:
     get:
-      summary: Get Cardano transactions set information
+      summary: Get Cardano transactions commitment information
```
Comment (jpraynaud) [inline openapi.yaml:291]:
```suggestion
      summary: Get Cardano transactions set commitment information
```

File: openapi.yaml [hunk 18]
```diff
@@ -267,52 +267,52 @@ paths:
...
       description: |
-        Returns the information of a Cardano transactions set
+        Returns the information of a Cardano transactions commitment
```
Comment (jpraynaud) [inline openapi.yaml:293]:
```suggestion
        Returns the information of a Cardano transactions set commitment
```

File: openapi.yaml [hunk 19]
```diff
@@ -267,52 +267,52 @@ paths:
...
       parameters:
         - name: hash
           in: path
-          description: Hash of the Cardano transactions set to retrieve
+          description: Hash of the Cardano transactions commitment to retrieve
```
Comment (jpraynaud) [inline openapi.yaml:297]:
```suggestion
          description: Hash of the Cardano transactions set commitment to retrieve
```

File: openapi.yaml [hunk 20]
```diff
@@ -267,52 +267,52 @@ paths:
...
           required: true
           schema:
             type: string
             format: bytes
           example: "6da2b104ed68481ef829d72d72c2f6a20142916d17985e01774b14ed49f0fea1"
       responses:
         "200":
-          description: Cardano transactions set found
+          description: Cardano transactions commitment found
```
Comment (jpraynaud) [inline openapi.yaml:305]:
```suggestion
          description: Cardano transactions set commitment found
```

File: openapi.yaml [hunk 21]
```diff
@@ -267,52 +267,52 @@ paths:
...
           content:
             application/json:
               schema:
-                $ref: "#/components/schemas/CardanoTransactionMessage"
+                $ref: "#/components/schemas/CardanoTransactionCommitmentMessage"
         "404":
-          description: Cardano transactions set not found
+          description: Cardano transactions commitment not found
```
Comment (jpraynaud) [inline openapi.yaml:311]:
```suggestion
          description: Cardano transactions set commitment not found
```

File: openapi.yaml [hunk 22]
```diff
@@ -267,52 +267,52 @@ paths:
...
         "412":
           description: API version mismatch
         default:
-          description: Cardano transactions set retrieval error
+          description: Cardano transactions commitment retrieval error
```
Comment (jpraynaud) [inline openapi.yaml:315]:
```suggestion
          description: Cardano transactions set commitment retrieval error
```

**Review by jpraynaud — APPROVED** [review]
LGTM :rocket: 

Very good work! 
Just few typos to fix before merging.

**Pull request merged**